Logging/LogHandler/FileLogHandler.cs

```diff
@@ -165,52 +176,107 @@ protected override void DisposeSelf()
                 callerDerivedTypeName, filename, callerName, callerLine
             );
 
-            m_Writer.WriteLine($"{timestamp}{logLevel}{context}{message}");
+            string log = $"{timestamp}{logLevel}{context}{message}";
 
-            if (m_ShouldRotate && m_Writer.BaseStream.Length > RotateFileSizeLimit)
+            if (m_RotateFileSizeLimit != null)
             {
-                this.RotateFiles();
+                int logLength = Encoding.Unicode.GetByteCount(log);
+                long fileLength = m_Writer.BaseStream.Length;
+
+                if ((fileLength + logLength) > m_RotateFileSizeLimit)
+                {
+                    if (logLength > m_RotateFileSizeLimit)
+                    {
+                        throw new Exception("Message size exceeds file size limit, logging failed " +
+                            $"({logLength} > {m_RotateFileSizeLimit})");
+                    }
+
+                    this.RotateFiles();
+                }
             }
+
+            m_Writer.WriteLine(log);
         }
 
         private void RotateFiles()
         {
-            Debug.Assert(RotateFileSizeLimit >= 0, $"Expected positive integer for file size limit, but got {RotateFileSizeLimit}");
-            Debug.Assert(RotateFileCountLimit >= 0, $"Expected positive integer for file count limit, but got {RotateFileCountLimit}");
-
-            m_Writer.Close();
+            m_Writer.Dispose();
             m_Writer = null;
 
-            // Determine how many rotated log files already exist
-            int index = 1;
-            string rotatedFilePath = GetRotatedFilePath(index);
-            while (File.Exists(rotatedFilePath))
+            Dictionary<int, string> fileMap = new Dictionary<int, string>{ [0] = m_Path };
+            foreach ((int index, string path) in GetRotatedFiles())
             {
-                rotatedFilePath = GetRotatedFilePath(++index);
+                if (index < 0 || index > m_RotateFileCountLimit)
+                {
+                    File.Delete(path);
+                }
+                else if (fileMap.ContainsKey(index))
+                {
+                    // In case of duplicate indices (ex. "log.1.txt" vs "log.001.txt") keep the shortest
+                    if (fileMap[index].Length < path.Length)
+                    {
+                        File.Delete(path);
+                    }
+                    else
+                    {
+                        File.Delete(fileMap[index]);
+                        fileMap[index] = path;
+                    }
+                }
+                else
+                {
+                    fileMap[index] = path;
+                }
             }
-            --index;
 
-            // Delete any files in excess of the file count limit
-            while (index >= RotateFileCountLimit)
+            RotateRecursive(0);
+
+            CreateWriter();
+
+            void RotateRecursive(int index)
```

Comment:
Tell me you've been writing a lot of TS/JS without telling me 😄

Reply:
I don't see it, what did I do? 😛

Reply:
inline functions. Much less frequently used in C# unless they're lambdas

Reply:
Oh, I used local functions long before I learned JS - any time I want to encapsulate logic relevant only to the current function, for use as a callback, LINQ predicate, recursion, etc.